src/components/swap/SwapBuyFiatButton.tsx

```diff
@@ -0,0 +1,149 @@
+    available: fiatOnrampAvailable,
+    availabilityChecked: fiatOnrampAvailabilityChecked,
+    error,
+    loading: fiatOnrampAvailabilityLoading,
+  } = useFiatOnrampAvailability(checkFiatRegionAvailability)
+  const [buyFiatFlowState, setBuyFiatFlowState] = useState(BuyFiatFlowState.INACTIVE)
+  const [walletDrawerOpen, toggleWalletDrawer] = useWalletDrawer()
+
+  /* 
+  Once the buy fiat flow has been completed, reset BuyFiatFlowState and also save to 
+  local storage that user has completed this flow already.
+  */
+  const terminateBuyFiatFlow = useCallback(() => {
+    setBuyFiatFlowCompleted(true)
+    setBuyFiatFlowState(BuyFiatFlowState.INACTIVE)
+  }, [setBuyFiatFlowCompleted, setBuyFiatFlowState])
+
+  /*
+  Depending on the current state of the buy fiat flow the user is in (buyFiatFlowState),
+  the desired behavior of clicking the 'Buy' button is different:
+  1) Initially upon first click, need to check the availability of the feature in the user's 
+  region, and continue the flow.
+  2) If the feature is available in the user's region, need to connect a wallet, and continue
+  the flow.
+  3) If the feature is available and a wallet account is connected, show fiat on ramp modal.
+  4) If the feature is unavailable, show feature unavailable tooltip. 
+  */
+  const handleBuyCrypto = useCallback(() => {
+    if (!fiatOnrampAvailabilityChecked) {
+      setCheckFiatRegionAvailability(true)
+      setBuyFiatFlowState(BuyFiatFlowState.ACTIVE_POST_REGION_CHECK)
+    } else if (fiatOnrampAvailable && !account && !walletDrawerOpen) {
+      toggleWalletDrawer()
+      setBuyFiatFlowState(BuyFiatFlowState.ACTIVE_NEEDS_WALLET_CONNECTION)
+    } else if (fiatOnrampAvailable && account) {
+      openFiatOnRampModal()
+      terminateBuyFiatFlow()
+    } else if (!fiatOnrampAvailable) {
+      terminateBuyFiatFlow()
+    }
+  }, [
+    fiatOnrampAvailabilityChecked,
+    fiatOnrampAvailable,
+    account,
+    walletDrawerOpen,
+    toggleWalletDrawer,
+    openFiatOnRampModal,
+    terminateBuyFiatFlow,
+  ])
+
+  // Continue buy fiat flow automatically when requisite state changes have occured.
+  useEffect(() => {
+    if (
+      buyFiatFlowState === BuyFiatFlowState.ACTIVE_POST_REGION_CHECK ||
+      (account && buyFiatFlowState === BuyFiatFlowState.ACTIVE_NEEDS_WALLET_CONNECTION)
+    ) {
+      handleBuyCrypto()
+    }
+  }, [account, handleBuyCrypto, buyFiatFlowState])
+
+  useEffect(() => {
+    if (error) console.error(error)
+  }, [error])
+
+  const disableBuyCryptoButton = Boolean(
+    (!fiatOnrampAvailable && fiatOnrampAvailabilityChecked) ||
+      fiatOnrampAvailabilityLoading ||
+      // When wallet drawer is open AND user is in the connect wallet step
+      // of the buy fiat flow, disable buy fiat button.
+      (walletDrawerOpen && buyFiatFlowState === BuyFiatFlowState.ACTIVE_NEEDS_WALLET_CONNECTION)
+  )
+
+  return (
+    <MouseoverTooltipContent
+      wrap
+      delayBeforeShow={POPOVER_DELAY_BEFORE_SHOW_MS}
```

Comment:
Why is this behavior necessary? It adds a lot of complexity to this PR.
Are we able to push back on design to remove this, for the sake of a simpler codebase?

Reply:
they specifically requested this so idts? unless @infredible ok's it - see thread convo here: https://uniswapteam.slack.com/archives/C04R3CV5887/p1680116729731119?thread_ts=1680116220.405679&cid=C04R3CV5887